Engine/media/audio/sound.cpp

```diff
@@ -53,31 +56,153 @@ static int GuessSoundTypeFromExt(const String &extension)
     return 0;
 }
 
+// Sound cache, stores most recent used sounds, tracks use history with MRU list.
+// TODO: refactor into the resource cache class, share with the sprite cache.
+class SoundCache
+{
+public:
+    typedef std::shared_ptr<std::vector<uint8_t>> DataRef;
+
+    void SetMaxCacheSize(size_t size)
+    {
+        FreeMem(size);
+        _maxSize = size;
+    }
+
+    DataRef Get(const String &name)
+    {
+        const auto found = _map.find(name);
+        if (found == _map.end())
+            return DataRef();
+        // Move to the beginning of the MRU list
+        _mru.splice(_mru.begin(), _mru, found->second.MruIt);
+        return found->second.Data;
+    }
+
+    // Add particular item into the cache, remove oldest items if cache size is exceeded
+    void Put(const String &name, DataRef &ref)
+    {
+        assert(ref);
+        if (!ref)
+            return; // safety precaution
+        if (Get(name))
+            return; // already in cache
+        if (_maxSize == 0)
+            return; // cache is disabled
+        // Clear up space before adding
+        if (_cacheSize + ref->size() > _maxSize)
+            FreeMem(_maxSize);
+        SoundEntry entry(name, ref);
+        entry.MruIt = _mru.insert(_mru.begin(), name);
+        _map[name] = std::move(entry);
+    }
+
+    // Clear the cache, dispose all resources
+    void Clear()
+    {
+        _map.clear();
+        _mru.clear();
+        _cacheSize = 0u;
+    }
+
+private:
+    // Delete the oldest (least recently used) item in cache
+    void DisposeOldest()
+    {
+        assert(_mru.size() > 0);
+        if (_mru.size() == 0) return;
+        auto it = std::prev(_mru.end());
+        const auto id = *it;
+        {
+            _cacheSize -= _map[id].Data->size();
+            _map[id].Data.reset();
+        }
+        // Remove from the mru list
+        _mru.erase(it);
+    }
+    // Keep disposing oldest elements until cache size is reduced to the given threshold
+    void FreeMem(size_t threshold)
```

Comment:
threshold is never used. Maybe `_cacheSize + threshold >= _maxSize` was supposed to be used later?